src/weights.jl

```diff
@@ -35,7 +40,33 @@ end
     W(v, sum(v))
 end
 
-Base.getindex(wv::W, ::Colon) where {W <: AbstractWeights} = W(copy(wv.values), sum(wv))
+Base.getindex(wv::AbstractWeights, ::Colon) = copy(wv)
+
+Base.copy(wv::W) where {W <: AbstractWeights} =
+    weightstype(W)(copy(wv.values), sum(wv))
+
+@propagate_inbounds function Base.view(wv::W, inds...) where
+    {S <: Real, W <: AbstractWeights{S}}
+    @boundscheck checkbounds(wv, inds...)
+    @inbounds v = invoke(view, Tuple{AbstractArray, Vararg{Any}}, wv, inds...)
+    weightstype(W){S, eltype(wv), typeof(v)}(v, missing)
+end
+
+# This method is implemented for backward compatibility
+@propagate_inbounds function Base.view(wv::W, inds::Union{Integer, CartesianIndex}...) where
+    {S <: Real, W <: AbstractWeights{S}}
+    @boundscheck checkbounds(wv, inds...)
+    @inbounds invoke(view, Tuple{AbstractArray, Vararg{Any}}, wv, inds...)
+end
+
+# Always recompute the sum for views of AbstractWeights, as we cannot know whether
```

Comment:
maybe move the definitions of `sum` and `copy` into one place to make sure the reader can see both definitions side by side?